Dev/adaq4224

machschmitt opened the pull request:
Add support for ADAQ4224 Data Acquisition micro module.

The ADAQ4224 is a micro module precision data acquisition (DAQ) signal chain solution.
The internal ADC is an AD4030-24 so this set leverages the driver that already supports AD4030-24 by expanding it to also support ADAQ4216, ADAQ4220, and ADAQ4224.

nunojsa commented on arch/arm/boot/dts/zynq-zed-adv7511-adaq4224-24.dts:
```diff
@@ -0,0 +1,148 @@
+			pwms = <&axi_pwm_gen 0 0>, <&axi_pwm_gen 1 0>;
+		};
+	};
+};
```
note: if you are carrying this patch on your PR, you could/should also sign off this patch

machschmitt replied:
ok, added sign off tag.

nunojsa commented on arch/arm/boot/dts/zynq-zed-adv7511-adaq4224-24_cm1_sdi4_cz2_ddr.dts:
```diff
@@ -0,0 +1,173 @@
+			pwms = <&axi_pwm_gen 0 0>, <&axi_pwm_gen 1 0>;
+		};
+	};
+};
```
Is hdl doing builds for these configurations? I'm not too keen in having devicetrees if we don't have HW for them...

machschmitt replied:
not sure, will checkout with the hw team. So, If they're not building with these configs I'll remove the related dts files.

nunojsa commented on drivers/iio/adc/ad4630.c:
```diff
@@ -1231,6 +1414,18 @@ static int ad4630_probe(struct spi_device *spi)
 	if (ret)
 		return ret;
 
+	if (st->chip->has_pga) {
```
I don't think the flag is that helpful in here... It's an optional GPIO so users won't add it if the device does not support it. Furthermore, we can make the bindings to enforce this

machschmitt replied:
that's right. Removed the check on V2.

nunojsa commented on drivers/iio/adc/ad4630.c:
```diff
@@ -1066,6 +1241,14 @@ static const struct iio_info ad4630_info = {
+static const struct iio_info adaq4224_info = {
+	.read_raw = &ad4630_read_raw,
+	.write_raw = &ad4630_write_raw,
+	.write_raw_get_fmt = &adaq4224_write_raw_get_fmt,
```
nit: don't mix function names... If these devices are going to be under this driver I would keep the same prefix for the functions. Anyways, up to you

machschmitt replied:
Sure. Changed on V2.

nunojsa commented on drivers/iio/adc/ad4630.c:
```diff
@@ -412,13 +490,74 @@ static int ad4630_set_chan_gain(struct iio_dev *indio_dev, int ch,
+
+static const struct attribute_group ad4630_attribute_group = {
+	.attrs = ad4630_attributes,
+};
```
Any reason to not use the typical attr_available flow?

machschmitt replied:
By typical attr_available you mean using info_mask_separate_available like [ad4130](https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/adc/ad4130.c#L962)?
I just saw it, will rework to use that.

nunojsa replied:
exactly

nunojsa commented on drivers/iio/adc/ad4630.c:
```diff
@@ -412,13 +490,74 @@ static int ad4630_set_chan_gain(struct iio_dev *indio_dev, int ch,
+		tmp0 = (int)div_s64_rem(tmp2, 1000000000LL, &tmp1);
+
+		len += scnprintf(buf + len, PAGE_SIZE - len,
+			"%d.%09u ", tmp0, abs(tmp1));
```
sysfs_emit_at()

It also seems to be me you could just have this computation once during probe and save the results... Or, am I missing something?

machschmitt replied:
Yes, it can. One thing I want to know first is whether developers should try to keep track of the current voltage reference any time it is used. For this driver (and I recall seeing others like this) it just stores the voltage ref on device probe and uses that throughout the device instance life cycle. Another way of doing it would be asking for the current voltage reference every time it is used for calculations. All regulator dt config I've done so far was with fixed-vref so doing the computation once at probe would be okay. I, however, haven't done many many vref configs so ...

nunojsa replied:
Yeah, it is very uncommon for something like vref to change. My approach is to just cache once during probe and then if that assumption changes, deal with it. Never changed so far...

nunojsa commented on drivers/iio/adc/ad4630.c:
```diff
@@ -386,6 +405,65 @@ static int ad4630_set_chan_offset(struct iio_dev *indio_dev, int ch, int offset)
+	int gain_idx;
+	u64 gain;
+
+	gain = gain_int * 1000000000 + gain_fract;
```
can you use units.h? ditto for the other places

machschmitt replied:
Used for the places where the constant was clearly related to some unit but avoided changing stuff that looks typical to IIO drivers such as `*val = (st->vref * 2) / 1000;`.

nunojsa commented on drivers/iio/adc/ad4630.c:
```diff
@@ -386,6 +405,65 @@ static int ad4630_set_chan_offset(struct iio_dev *indio_dev, int ch, int offset)
+
+static int ad4630_set_pga_gain(struct iio_dev *indio_dev, int gain_idx)
+{
+	unsigned long *values = bitmap_alloc(ADAQ42xx_PGA_PINS, GFP_KERNEL);
```
I would likely prefer to alloc the bitmap once during probe (when needed) rather that doing it every time when changing gains

machschmitt replied:
Changed it to use `DECLARE_BITMAP` so the bitmap stays in function stack memory.
Let me know if that's okay otherwise, I'll allocate the bitmap on probe as suggested.

nunojsa reviewed (commented):
Looks good... Some minor notes

machschmitt commented:
Alright, let's provide a proper change log.

Change log V1 -> V2
- Changed function declarations and defines to keep the same prefix for the functions
- Added my sign-off tag to all commits
- Removed if checks from probe function when possible
- Replaced `bitmap_alloc` with `DECLARE_BITMAP` so the values for GPIO states are stored in function stack memory instead of heap memory.
- Use constants from units.h to replace a few hardcoded values are were clearly related to SI values.
- Reworked voltage scale available attribute to use the typical IIO attr available control flow.
- Stored the computation of the scale available values on probe rather than recalculating them every call to in_voltage_scale_available.

Pending:
Need to check with the HDL team if they do builds for the dts files in this PR.

machschmitt commented:
Change log V2 -> V3
- Removed extra dts file the HDL team didn't ask to keep.
- dt-binding patch is now the first in the series
- Squased patches adding dts files

nunojsa commented on drivers/iio/adc/ad4630.c:
```diff
@@ -95,8 +99,16 @@
+#define ADAQ4224_PGA_1_BITMAP		0
+#define ADAQ4224_PGA_2_BITMAP		BIT(0)
+#define ADAQ4224_PGA_3_BITMAP		BIT(1)
+#define ADAQ4224_PGA_4_BITMAP		(BIT(1) | BIT(0))
```
GENMASK(1, 0)?

machschmitt replied:
done

nunojsa commented on drivers/iio/adc/ad4630.c:
```diff
@@ -95,8 +99,16 @@
-#define AD4630_VREF_MIN			(4096 * 1000)
-#define AD4630_VREF_MAX			(5000 * 1000)
+#define AD4630_VREF_MIN			(4096 * MILLI)
+#define AD4630_VREF_MAX			(5000 * MILLI)
```
a bit unrelated but alright

nunojsa reviewed (approved):
LGTM... 

I would still do the GENMASK() change but up to you...

machschmitt commented:
We made a minor fix to of_match table when testing the driver with ad4030.
For ad4030, there were also some dts changes. I'm assuming they apply to ad4630 but am also checking with @ladace.
If the change to `lane-mode` and other dt properties is only for ad4030 then we can just drop the last patch in the series.

nunojsa commented on drivers/iio/adc/ad4630.c:
```diff
@@ -1535,9 +1535,9 @@ static const struct spi_device_id ad4630_id_table[] = {
+	{ .compatible = "adi,ad4030-24", .data = &ad4630_chip_info[ID_AD4030_24] },
 	{ .compatible = "adi,ad4630-16", .data = &ad4630_chip_info[ID_AD4630_16] },
-	{ .compatible = "adi,ad4030-24", .data = &ad4630_chip_info[ID_AD4630_24] },
+	{ .compatible = "adi,ad4630-24", .data = &ad4630_chip_info[ID_AD4630_24] },
```
oh... doh!

Please, have this as the first patch in this series (or the first one touching the file). Then, we need to backport this at least to the 2022_R2 branch!

machschmitt commented:
Rebased to put new patches at the beginning of the series.
@ladace told me the dt updates (patch 2 now) do apply for ad4630 so better keep that.
